### ISSUE TITLE
Test regression with pytest-8.0.0: `tests/test_exceptions.py::test_response_body[RequestRedirect]`

### ISSUE THREAD

**Issue by mgorny** [pallets/werkzeug#2845]
After upgrading to `pytest-8.0.0`, a single test fails:

```pytb
$ python -m pytest
========================================================= test session starts =========================================================
platform linux -- Python 3.11.7, pytest-8.0.0, pluggy-1.3.0
rootdir: /tmp/werkzeug
configfile: pyproject.toml
testpaths: tests
plugins: xprocess-0.23.0, timeout-2.2.0
collected 897 items                                                                                                                   

tests/middleware/test_dispatcher.py .                                                                                           [  0%]
tests/middleware/test_http_proxy.py .                                                                                           [  0%]
tests/middleware/test_lint.py .........                                                                                         [  1%]
tests/middleware/test_profiler.py .                                                                                             [  1%]
tests/middleware/test_proxy_fix.py .................                                                                            [  3%]
tests/middleware/test_shared_data.py ..                                                                                         [  3%]
tests/sansio/test_multipart.py .....................                                                                            [  5%]
tests/sansio/test_request.py .........                                                                                          [  6%]
tests/sansio/test_utils.py ....................                                                                                 [  9%]
tests/test_datastructures.py .................................................................................................. [ 19%]
                                                                                                                                [ 19%]
tests/test_debug.py .....................                                                                                       [ 22%]
tests/test_exceptions.py .........................................................F............................                 [ 31%]
tests/test_formparser.py .........................                                                                              [ 34%]
tests/test_http.py ............................................................................................................ [ 46%]
......                                                                                                                          [ 47%]
tests/test_internal.py .                                                                                                        [ 47%]
tests/test_local.py ...............................................................                                             [ 54%]
tests/test_routing.py ......................................................................................................... [ 66%]
............                                                                                                                    [ 67%]
tests/test_security.py .........                                                                                                [ 68%]
tests/test_send_file.py ................................                                                                        [ 72%]
tests/test_serving.py ............s.............                                                                                [ 75%]
tests/test_test.py ...........................................................                                                  [ 81%]
tests/test_urls.py ................                                                                                             [ 83%]
tests/test_utils.py ....................................                                                                        [ 87%]
tests/test_wrappers.py ............................................................................................             [ 97%]
tests/test_wsgi.py .....................                                                                                        [100%]

============================================================== FAILURES ===============================================================
_________________________________________________ test_response_body[RequestRedirect] _________________________________________________

cls = <class 'werkzeug.routing.exceptions.RequestRedirect'>

    @pytest.mark.parametrize(
        "cls",
        sorted(
            (e for e in HTTPException.__subclasses__() if e.code),
            key=lambda e: e.code,  # type: ignore
        ),
    )
    def test_response_body(cls):
>       exc = cls()
E       TypeError: RequestRedirect.__init__() missing 1 required positional argument: 'new_url'

tests/test_exceptions.py:166: TypeError
======================================================= short test summary info =======================================================
FAILED tests/test_exceptions.py::test_response_body[RequestRedirect] - TypeError: RequestRedirect.__init__() missing 1 required positional argument: 'new_url'
============================================== 1 failed, 895 passed, 1 skipped in 16.55s ==============================================
```

Confirmed on d3dd65a27388fbd39d146caacf2563639ba622f0.

I've reproduced via:

```
tox -e py311
. .tox/py311/bin/activate
pip install -U pytest
python -m pytest
```

Environment:

- Python version: 3.11.7
- Werkzeug version: 3.0.1, d3dd65a27388fbd39d146caacf2563639ba622f0

**Comment by mgorny** [comment]
To copy some context that I've added to #2846:

The test runs over `HTTPException.__subclasses__()`. Most of these subclasses are defined in `werkzeug.exceptions`, and then there's a `RequestRedirect` class in `werkzeug.routing.exceptions`. Importing these modules affect what `__subclasses__()` contain, though I don't know if GC affects that somehow as well.

Now, with `pytest < 8` apparently `RequestRedirect` wasn't in scope, and therefore wasn't tested. In `pytest >= 8`, it started being in scope and since the test didn't account for it, it failed. Though it's pretty much undefined whether to expect it or not.

**Comment by pgjones** [comment]
Fixed by 4e5bdca7f8227d10cae828f8064fb98190ace4aa
